File: apps/ios/PassAlarm/Domain/UseCase/CheckProLimitUseCase.swift
import Foundation

final class CheckProLimitUseCase: Sendable {
    private let planRepository: AlarmPlanRepositoryProtocol
    private let subscriptionRepository: SubscriptionRepositoryProtocol

    private static let freePlanLimit = 10
    private static let proPlanLimit = 100

    init(planRepository: AlarmPlanRepositoryProtocol,
         subscriptionRepository: SubscriptionRepositoryProtocol) {
        self.planRepository = planRepository
        self.subscriptionRepository = subscriptionRepository
    }

    /// Returns `true` if the user can create a new alarm plan within their tier limit.
    func execute() async throws -> Bool {
        let status = await subscriptionRepository.currentStatus()
        let plans = try await planRepository.fetchAll()
        let limit = status.isPro ? Self.proPlanLimit : Self.freePlanLimit
        return plans.count < limit
    }
}

File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/usecase/CheckProLimitUseCase.kt
package com.nogeass.passalarm.domain.usecase

import com.nogeass.passalarm.domain.repository.AlarmPlanRepository
import com.nogeass.passalarm.domain.repository.SubscriptionRepository
import javax.inject.Inject

class CheckProLimitUseCase @Inject constructor(
    private val planRepository: AlarmPlanRepository,
    private val subscriptionRepository: SubscriptionRepository,
) {
    companion object {
        private const val FREE_PLAN_LIMIT = 10
        private const val PRO_PLAN_LIMIT = 100
    }

    suspend fun canCreatePlan(): Boolean {
        val plans = planRepository.fetchAll()
        val status = subscriptionRepository.currentStatus()
        val limit = if (status.isPro) PRO_PLAN_LIMIT else FREE_PLAN_LIMIT
        return plans.size < limit
    }
}
